### PR TITLE
Add passportJs

### LINKED ISSUE #26: Use passportjs for auth
passport-local should do

### PR DESCRIPTION
**Pull request by unsatisfiedpopcorn**
Resolves #26 

# Rider route has been updated to conform to new authentication flow

@remo5000 Please review and note the rest of the routes that need updating, then split the workload between @chikchengyao  and @evanmok2401 estimated completion by Tuesday?

Should not be a lot of work. Mainly need to find instances where userids are used then change them to take their inputs from req.user instead.

### Moving forward:
- What should the default page be after a successful login?
- How should the users page work after the introduction of authentication?

@remo5000 I can set the default page since I wrote the redirection. Please distribute the work on the users page evenly / start a discussion if needed.

### CODE REVIEW

File: routes/rider.js [hunk 1]
```diff
@@ -1,10 +1,11 @@
 var express = require('express')
 var router = express.Router()
 var db = require('../model/db.js')
+const connect = require('connect-ensure-login')
 
 /* GET user-rider's rides/bids listing page. */
-router.get('/', async function(req, res, next) {
-  const riderId = req.query.user_id_field
+router.get('/', connect.ensureLoggedIn('/login') ,async function(req, res, next) {
```
Comment (remo5000) [inline routes/rider.js]: Is it possible to remove the /login here since the other routers don't have them?

File: routes/rider.js [hunk 2]
```diff
@@ -1,10 +1,11 @@
...
+  const riderId = req.user
```
Comment (remo5000) [inline routes/rider.js]: Is there a way to set this as `userId` for passport? Because I'm getting confused between username and userid (especially with no types 😆)

File: routes/rider.js (hunk 1 above)
Comment (unsatisfiedpopcorn) [reply routes/rider.js]: This is ok, connect.ensureLoggedIn('/login') redirects to `domain/login` which is the login page. it does not redirect to `domain/route/login`

File: config/passport.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+passport.use(new LocalStrategy(
+    async (username, password, done) => {
+      const userData = await db.user.getValidUserid(username, password, db.exposedInstance)
+      const unwrappedUserData = userData[0]
+      console.log("JSON OBJECT: " + JSON.stringify(unwrappedUserData))
+
+      if (unwrappedUserData) {
+        return done(null, unwrappedUserData.userid)
+      } else {
+        return done(userData, false)
```
[the remark this answers is not in the document]
Comment (unsatisfiedpopcorn) [reply config/passport.js]: `userData` could be an err bubbled up.  sorry bad naming/ lack of comments

File: routes/rider.js (hunk 2 above)
Comment (unsatisfiedpopcorn) [reply routes/rider.js]: Did some quick google-fu, not to my knowledge sorry. It's not as simple as writing `req.userId = req.user` ; `req.user = undefined`; i think due to the fact that `req.user` is required as the default for some methods in passportjs. 

tl;dr seems like there should be a way but i'm not sure

File: model/user.js [hunk 4]
```diff
@@ -102,9 +102,32 @@ const listCarUserOwns = async (owner, db) => {
...
+      FROM appUserAccount a
+      WHERE 
+        a.email = $1 AND
+        a.password = $2
+      `
+      ,
+      [username, password]
+  )
+  .then(result => {
+    console.log(`Successful login for userID: ${JSON.stringify(result)}`)
```
[the remark this answers is not in the document]
Comment (unsatisfiedpopcorn) [reply model/user.js]: Sure, will reference as an issue. I'll fix it the next time I work on this,
